core/registration.c

```diff
@@ -840,23 +877,36 @@ static int prv_updateRegistration(lwm2m_context_t * contextP,
             lwm2m_free(payload);
             return COAP_500_INTERNAL_SERVER_ERROR;
         }
-        coap_set_payload(transaction->message, payload, payload_length);
+
+        transaction_set_payload(transaction, payload, payload_length);
     }
 
+    registration_data_t * dataP = lwm2m_malloc(sizeof(registration_data_t));
+    if (dataP == NULL){
+        lwm2m_free(payload);
+        return COAP_500_INTERNAL_SERVER_ERROR;
+    }
+
+    dataP->payload = payload;
+    dataP->server = server;
+
     transaction->callback = prv_handleRegistrationUpdateReply;
-    transaction->userData = (void *) server;
+    transaction->userData = (void *) dataP;
 
     contextP->transactionList = (lwm2m_transaction_t *)LWM2M_LIST_ADD(contextP->transactionList, transaction);
 
     if (transaction_send(contextP, transaction) == 0)
     {
         server->status = STATE_REG_UPDATE_PENDING;
     }
-
-    if (withObjects == true)
-    {
-        lwm2m_free(payload);
+    else {
+        if (withObjects == true)
+        {
+            lwm2m_free(payload);
```

Comment:
Here in case of the registration update with objects included, we will free the payload, but not set it to NULL in the dataP. This leads to double-free in prv_handleRegistrationUpdateReply

Reply:
Thank you, I currently looking into some other (possible) memory issues and will have a look at this as well.